File: tic-tac-toe-core/src/main/java/by/xxx/pupil/ai/minimax/findmoves/ThreatMovesFinder.java
package by.xxx.pupil.ai.minimax.findmoves;

import by.xxx.pupil.CombinationsFinder;
import by.xxx.pupil.GeneralCombination;
import by.xxx.pupil.model.Board;
import by.xxx.pupil.model.CellType;
import by.xxx.pupil.model.Move;
import by.xxx.pupil.model.Player;
import org.apache.commons.lang3.Validate;

import java.util.ArrayList;
import java.util.List;

public class ThreatMovesFinder implements MovesFinder {

    private final CombinationsFinder combinationsFinder;
    private final InRadiusMovesFinder inRadiusMovesFinder;

    public ThreatMovesFinder(
            InRadiusMovesFinder inRadiusMovesFinder,
            CombinationsFinder combinationsFinder
    ) {
        Validate.notNull(combinationsFinder, "combinationsFinder is null");
        Validate.notNull(inRadiusMovesFinder, "defaultMovesFinder is null");

        this.combinationsFinder = combinationsFinder;
        this.inRadiusMovesFinder = inRadiusMovesFinder;
    }

    @Override
    public List<Move> getMoves(Board board, Player player) {
        List<Move> moves = inRadiusMovesFinder.getMoves(board, player);

        List<Move> threatSpaceMoves = new ArrayList<>();

        for (Move move : moves) {
            board.updateCellValue(move.getI(), move.getJ(), Player.CROSSES.equals(player) ? CellType.CROSS : CellType.NOUGHT);
            List<GeneralCombination> patternsList = combinationsFinder.getCombinations(board, move);
            if (
                    patternsList.contains(GeneralCombination.STRAIGHT_FOUR) ||
                            patternsList.contains(GeneralCombination.FOUR) ||
                            patternsList.contains(GeneralCombination.THREE) ||
                            patternsList.contains(GeneralCombination.BROKEN_THREE)
            ) {
                threatSpaceMoves.add(move);
            }
            board.updateCellValue(move.getI(), move.getJ(), CellType.EMPTY);
        }

        return !threatSpaceMoves.isEmpty() ? threatSpaceMoves : moves;
    }
}

File: tic-tac-toe-core/src/main/kotlin/by/xxx/pupil/ai/minimax/findmoves/ThreatMovesFinder.kt
package by.xxx.pupil.ai.minimax.findmoves

import by.xxx.pupil.ai.combinations.CombinationsFinder
import by.xxx.pupil.ai.combinations.GeneralCombination
import by.xxx.pupil.model.Board
import by.xxx.pupil.model.CellType
import by.xxx.pupil.model.Move
import by.xxx.pupil.model.Player

class ThreatMovesFinder(
        private val inRadiusMovesFinder: InRadiusMovesFinder,
        private val combinationsFinder: CombinationsFinder
) : MovesFinder {

    override fun getMoves(board: Board, player: Player): List<Move> {
        val moves = inRadiusMovesFinder.getMoves(board, player)
        val threatSpaceMoves: MutableList<Move> = ArrayList()
        for (move in moves) {
            board.updateCellValue(move.y, move.x, if (Player.CROSSES == player) CellType.CROSS else CellType.NOUGHT)
            val patternsList = combinationsFinder.getCombinations(board, move)
            if (patternsList.contains(GeneralCombination.STRAIGHT_FOUR) ||
                    patternsList.contains(GeneralCombination.FOUR) ||
                    patternsList.contains(GeneralCombination.THREE) ||
                    patternsList.contains(GeneralCombination.BROKEN_THREE)) {
                threatSpaceMoves.add(move)
            }
            board.updateCellValue(move.y, move.x, CellType.EMPTY)
        }
        return if (!threatSpaceMoves.isEmpty()) threatSpaceMoves else moves
    }

}